src/goto-programs/remove_exceptions.cpp

```diff
@@ -98,175 +101,145 @@ class remove_exceptionst
     goto_programt &goto_program,
     const goto_programt::targett &instr_it,
     const stack_catcht &stack_catch,
     const std::vector<exprt> &locals);
 
   void instrument_throw(
-    const irep_idt &function_id,
     goto_programt &goto_program,
     const goto_programt::targett &,
     const stack_catcht &,
     const std::vector<exprt> &);
 
   void instrument_function_call(
-    const irep_idt &function_id,
     goto_programt &goto_program,
     const goto_programt::targett &,
     const stack_catcht &,
     const std::vector<exprt> &);
 
   void instrument_exceptions(
-    const irep_idt &function_id,
     goto_programt &goto_program);
 };
 
-/// adds exceptional return variables for every function that may escape
-/// exceptions
-void remove_exceptionst::add_exceptional_returns(
-  const irep_idt &function_id,
-  goto_programt &goto_program)
+/// Create a global named java::@inflight_exception that holds any exception
+/// that has been thrown but not yet caught.
+symbol_exprt remove_exceptionst::get_or_create_inflight_exception_global()
 {
+  const symbolt *existing_symbol =
+    symbol_table.lookup(INFLIGHT_EXCEPTION_VARIABLE_NAME);
+  if(existing_symbol)
+    return existing_symbol->symbol_expr();
+
+  auxiliary_symbolt new_symbol;
+  new_symbol.is_static_lifetime = true;
+  new_symbol.base_name = "@inflight_exception";
+  new_symbol.name = INFLIGHT_EXCEPTION_VARIABLE_NAME;
+  new_symbol.mode = ID_java;
+  new_symbol.type = pointer_type(empty_typet());
+  new_symbol.is_file_local = false;
+
+  symbolt *ret;
+  bool move_failed = symbol_table.move(new_symbol, ret);
+  INVARIANT(!move_failed, "inflight exception symbol should not be in table");
+
+  return ret->symbol_expr();
+}
 
-  auto maybe_symbol=symbol_table.lookup(function_id);
-  INVARIANT(maybe_symbol, "functions should be recorded in the symbol table");
-  const symbolt &function_symbol=*maybe_symbol;
-
-  // for now only add exceptional returns for Java
-  if(function_symbol.mode!=ID_java)
-    return;
-
-  if(goto_program.empty())
-    return;
-
-  // some methods (e.g. the entry method) have already been added to
-  // the symbol table; if you find it, initialise it
-  maybe_symbol=symbol_table.lookup(id2string(function_id)+EXC_SUFFIX);
-  if(maybe_symbol)
-  {
-    const symbolt &symbol=*maybe_symbol;
-    symbol_exprt lhs_expr_null=symbol.symbol_expr();
-    null_pointer_exprt rhs_expr_null((pointer_type(empty_typet())));
-    goto_programt::targett t_null=
-      goto_program.insert_before(goto_program.instructions.begin());
-    t_null->make_assignment();
-    t_null->source_location=
-      goto_program.instructions.begin()->source_location;
-    t_null->code=code_assignt(
-      lhs_expr_null,
-      rhs_expr_null);
-    t_null->function=function_id;
-    return;
-  }
+/// Adds a null initializer for the global @inflight_exception
+/// \param initialize_function: program to add the initializer to
+void remove_exceptionst::initialize_inflight_exception_global(
```

Comment:
Should the symbol just always be part of the symbol table (at least for Java, and possibly also C++), and hence be correctly initialized automatically?